parser/src/ast/expression.rs

```diff
@@ -0,0 +1,719 @@
+    Vector(Span<Vec<ScalarExpr>>),
+    /// A matrix of scalar expressions
+    Matrix(Span<Vec<Vec<ScalarExpr>>>),
+    /// A generator expression which produces a vector or matrix of values
+    ListComprehension(ListComprehension),
+}
+impl Expr {
+    /// Returns true if this expression is constant
+    ///
+    /// NOTE: This only returns true for the `Const` and `Range` variants
+    pub fn is_constant(&self) -> bool {
+        matches!(self, Self::Const(_) | Self::Range(_))
+    }
+}
+impl fmt::Debug for Expr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::Const(ref expr) => f.debug_tuple("Const").field(expr).finish(),
+            Self::Scalar(ref expr) => f.debug_tuple("Scalar").field(expr).finish(),
+            Self::Range(ref expr) => f.debug_tuple("Range").field(&expr.item).finish(),
+            Self::Vector(ref expr) => f.debug_tuple("Vector").field(&expr.item).finish(),
+            Self::Matrix(ref expr) => f.debug_tuple("Matrix").field(&expr.item).finish(),
+            Self::ListComprehension(ref expr) => {
+                f.debug_tuple("ListComprehension").field(expr).finish()
+            }
+        }
+    }
+}
+impl fmt::Display for Expr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::Const(ref expr) => write!(f, "{}", expr.ty()),
+            Self::Scalar(ref expr) => write!(f, "{}", expr),
+            Self::Range(_) => f.write_str("range"),
+            Self::Vector(ref expr) => write!(f, "vector of length {}", expr.len()),
+            Self::Matrix(ref expr) => write!(
+                f,
+                "matrix of {} rows and {} columns",
+                expr.len(),
+                expr.first().unwrap().len()
+            ),
+            Self::ListComprehension(_) => write!(f, "list comprehension"),
+        }
+    }
+}
+impl From<ScalarExpr> for Expr {
+    #[inline]
+    fn from(expr: ScalarExpr) -> Self {
+        Self::Scalar(expr)
+    }
+}
+impl From<ListComprehension> for Expr {
+    #[inline]
+    fn from(expr: ListComprehension) -> Self {
+        Self::ListComprehension(expr)
+    }
+}
+
+/// Scalar expressions are expressions which evaluate to a single scalar value,
+/// i.e. they have no vector or matrix elements. Only scalar expressions are valid
+/// in a constraint statement.
+#[derive(Clone, PartialEq, Eq, Spanned)]
+pub enum ScalarExpr {
+    /// A constant scalar value, i.e. integer
+    Const(Span<u64>),
+    /// A reference to a named value that must produce a scalar
+    SymbolAccess(SymbolAccess),
+    /// A reference to a trace column on a particular boundary of the trace, which must produce a scalar
+    ///
+    /// NOTE: This is only a valid expression in boundary constraints
+    BoundedSymbolAccess(BoundedSymbolAccess),
+    /// A binary operator over scalar values
+    Binary(BinaryExpr),
+    /// A call to a function or evaluator
+    Call(Call),
```

Comment:
Sure thing, I'll flesh out the docs on the `Call` type to clarify why it has the type field at all, as well as why it is an `Option`

Reply:
The docs for `Call` have been updated, let me know if you think that helps clarify things!

Reply:
Thanks, I think the docs for `Call` are clear. I would still prefer to have a note here, though, similar to how you put one above `BoundedSymbolAccess`, because the docs for this say "Scalar expressions are expressions which evaluate to a single scalar value", but evaluator functions don't return a scalar value. I don't think it needs to be anything detailed since there are detailed docs elsewhere, but it would be good to have something for times when this is read independently of the `Call` docs.